src/types/quota.rs

```diff
@@ -0,0 +1,160 @@
+            QuotaLimitName::Message => write!(f, "MESSAGE"),
+            QuotaLimitName::Atom(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+#[self_referencing]
+pub struct Quota {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) quota: imap_proto::Quota<'this>,
+}
+
+impl Quota {
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        QuotaTryBuilder {
+            data: owned,
+            quota_builder: |input| {
+                let mut lines: &[u8] = input;
+
+                // There should only be one single QUOTA response
+                while !lines.is_empty() {
+                    match imap_proto::parser::parse_response(lines) {
+                        Ok((_rest, Response::Quota(a))) => {
+                            return Ok(a);
+                        }
+                        Ok((rest, data)) => {
+                            lines = rest;
+                            if let Some(resp) = try_handle_unilateral(data, unsolicited) {
+                                return Err(resp.into());
+                            }
+                        }
+                        _ => {
+                            return Err(Error::Parse(ParseError::Invalid(lines.to_vec())));
+                        }
+                    }
+                }
+
+                Err(Error::Parse(ParseError::Invalid(lines.to_vec())))
+            },
+        }
+        .try_build()
+    }
+
+    pub fn root_name(&self) -> &str {
+        &*self.borrow_quota().root_name
+    }
+
+    pub fn resources(&self) -> &[imap_proto::QuotaResource<'_>] {
+        &*self.borrow_quota().resources
+    }
+}
+
+#[self_referencing]
+pub struct QuotaRoot {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) inner: InnerQuotaRoot<'this>,
+}
+
+pub(crate) struct InnerQuotaRoot<'a> {
+    pub(crate) quota_root: imap_proto::QuotaRoot<'a>,
+    pub(crate) quotas: Vec<imap_proto::Quota<'a>>,
+}
+
+impl QuotaRoot {
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        QuotaRootTryBuilder {
+            data: owned,
+            inner_builder: |input| {
+                let mut quota_roots: Vec<imap_proto::QuotaRoot<'_>> = Vec::new();
+                let mut quotas = Vec::new();
+                parse_many_into(
+                    input,
+                    &mut quota_roots,
+                    unsolicited,
+                    |response| match response {
+                        Response::QuotaRoot(q) => Ok(MapOrNot::Map(q)),
+                        Response::Quota(_) => Ok(MapOrNot::Ignore),
+                        resp => Ok(MapOrNot::Not(resp)),
+                    },
+                )?;
+                parse_many_into(input, &mut quotas, unsolicited, |response| match response {
```

Comment:
I _think_ this will have the effect of sending any unsolicited items in `input` twice. We may need to make `parse_many_into` able to get a `None` for `unsolicited` so that for the second pass over the input ignores unsolicited things. Or I _think_ you could maybe do it by just mapping anything that's not `::Quota` to `::Ignore`.

Reply:
Iterating over the input twice to parse it is somewhat unfortunate, though I think it's okay. It means we avoid duplicating the parsing loop, and quota replies are likely to be fairly short and not performance sensitive anyway.

Reply:
yeah..  I didn't like it..  I have reworked this now by creating a parse_many_into2 function that iterates once.